source/simulator/stokes_matrix_free.cc

```diff
@@ -1606,6 +1674,126 @@ namespace aspect
         mg_matrices_Schur_complement[level].fill_cell_data (level_viscosity_tables[level],
                                                             sim.pressure_scaling);
       }
+
+    {
+      // create active mesh tables for derivatives needed in Newton method
+      // and the strain rate.
+      if (sim.newton_handler && sim.newton_handler->parameters.newton_derivative_scaling_factor!=0)
+        {
+          const double newton_derivative_scaling_factor =
+            sim.newton_handler->parameters.newton_derivative_scaling_factor;
+
+          FEValues<dim> fe_values (*sim.mapping,
+                                   sim.finite_element,
+                                   quadrature_formula,
+                                   update_values   |
+                                   update_gradients |
+                                   update_quadrature_points |
+                                   update_JxW_values);
+
+          MaterialModel::MaterialModelInputs<dim> in(fe_values.n_quadrature_points, sim.introspection.n_compositional_fields);
+          MaterialModel::MaterialModelOutputs<dim> out(fe_values.n_quadrature_points, sim.introspection.n_compositional_fields);
+          sim.newton_handler->create_material_model_outputs(out);
+#if DEAL_II_VERSION_GTE(9,3,0)
+          const unsigned int n_cells = stokes_matrix.get_matrix_free()->n_cell_batches();
+#else
+          const unsigned int n_cells = stokes_matrix.get_matrix_free()->n_macro_cells();
+#endif
+          const unsigned int n_q_points = quadrature_formula.size();
+
+          active_strain_rate_table.reinit(TableIndices<2>(n_cells, n_q_points));
+          active_viscosity_derivative_wrt_pressure_table.reinit(TableIndices<2>(n_cells, n_q_points));
+          active_viscosity_derivative_wrt_strain_rate_table.reinit(TableIndices<2>(n_cells, n_q_points));
+
+          for (unsigned int cell=0; cell<n_cells; ++cell)
+            {
+#if DEAL_II_VERSION_GTE(9,3,0)
+              const unsigned int n_components_filled = stokes_matrix.get_matrix_free()->n_active_entries_per_cell_batch(cell);
+#else
+              const unsigned int n_components_filled = stokes_matrix.get_matrix_free()->n_components_filled(cell);
+#endif
+              for (unsigned int i=0; i<n_components_filled; ++i)
+                {
+                  typename DoFHandler<dim>::active_cell_iterator FEQ_cell =
+                    stokes_matrix.get_matrix_free()->get_cell_iterator(cell,i);
+                  typename DoFHandler<dim>::active_cell_iterator sim_cell(&(sim.triangulation),
+                                                                          FEQ_cell->level(),
+                                                                          FEQ_cell->index(),
+                                                                          &(sim.dof_handler));
```

Comment:
What is the difference between `FEQ_cell` and `sim_cell`? Leave a comment above why you need both. Also the names are not explaining their purpose. I would guess better names could be `matrix_free_stokes_cell` and `simulator_cell`?